convert/convert.go

```diff
@@ -0,0 +1,120 @@
+package convert
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kong/deck/file"
+	"github.com/kong/deck/utils"
+	"github.com/kong/go-kong/kong"
+)
+
+type Format string
+
+const (
+	FormatKongGateway Format = "kong-gateway"
+	FormatKonnect     Format = "konnect"
+)
+
+var (
+	AllFormats = []Format{FormatKongGateway, FormatKonnect}
+
+	conversions = map[Format]map[Format]bool{
+		FormatKongGateway: {
+			FormatKonnect: true,
+		},
+	}
+)
+
+func validConversion(from, to Format) (bool, error) {
+	if fromMap, ok := conversions[from]; ok {
+		if _, ok := fromMap[to]; ok {
+			return ok, nil
+		}
+	}
+
+	return false, fmt.Errorf("cannot convert from '%s' to '%s' format", from, to)
+}
+
+func ParseFormat(key string) (Format, error) {
+	format := Format(strings.ToLower(key))
+	switch format {
+	case FormatKongGateway:
+		return FormatKongGateway, nil
+	case FormatKonnect:
+		return FormatKonnect, nil
+	default:
+		return "", fmt.Errorf("invalid format: '%v'", key)
+	}
+}
+
+func Convert(inputFilename, outputFilename string, from, to Format) error {
+	if valid, err := validConversion(from, to); !valid {
+		return err
+	}
+
+	switch {
+	case from == FormatKongGateway && to == FormatKonnect:
+		return convertKongGatewayToKonnect(inputFilename, outputFilename)
+	default:
+		return fmt.Errorf("cannot convert from '%s' to '%s' format", from, to)
+	}
```

Comment:
We're passing `inputFilename` and `outputFilename` down the function calls several times without a clear reason, causing `convertKongGatewayToKonnect` to be concerned not only with conversion logic, but also with file handling.

Also, with more conversion format pairs added, file handling logic will need to be copied across all these format pair implementations.

How about doing file operations in `Convert` and having conversions as simple functional transformations (without the need to bloat actual conversion logic with the file handling concern)

```go
func Convert(inputFilename, outputFilename string, from, to Format) error {
  // probably unnecessary - see my comment above
	if valid, err := validConversion(from, to); !valid {
		return err
	}

	inputContent, err := file.GetContentFromFiles([]string{inputFilename})
	if err != nil {
		return err
	}

	var outputContent *file.Content
	switch {
	case from == FormatKongGateway && to == FormatKonnect:
		outputContent = convertKongGatewayToKonnect(inputContent)
	default:
		return fmt.Errorf("cannot convert from '%s' to '%s' format", from, to)
	}

	return file.WriteContentToFile(outputContent, outputFilename, file.YAML)
}
```